### PR TITLE
Add package: LuaFormatter

### PR DESCRIPTION
**Pull request by shaeinst**

### CODE REVIEW

File: lua/mason-registry/luaformatter/init.lua [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+return Pkg.new {
+    name = "luaformatter",
+    desc = [[Code formatter for Lua]],
+    homepage = "https://github.com/Koihik/LuaFormatter",
+    languages = { Pkg.Lang.Lua },
+    categories = { Pkg.Cat.Formatter },
+    ---@async
+    ---@param ctx InstallContext
+    install = function(ctx)
```
Comment (williamboman) [inline lua/mason-registry/luaformatter/init.lua:15]: It seems they distribute via luarocks, which they also recommend. I think we should use that method to install (there's already a luarocks manager, might need to be slightly extended though)

Comment (shaeinst) [reply lua/mason-registry/luaformatter/init.lua:15]: i actually tried it implementing with luarocks but i couldn't get it working. they also seems didn't update dev branch in a while. so, i thought going with c(make) would be good choice.

**Comment by williamboman** [comment]
@williambotman /apply

```diff
diff --git a/lua/mason-core/managers/luarocks/init.lua b/lua/mason-core/managers/luarocks/init.lua
index 0deb2e1..9fd59ca 100644
--- a/lua/mason-core/managers/luarocks/init.lua
+++ b/lua/mason-core/managers/luarocks/init.lua
@@ -21,7 +21,7 @@ local function with_receipt(package)
 end
 
 ---@param package string: The luarock package to install.
----@param opts { dev: boolean, bin : string[] | nil } | nil
+---@param opts { dev: boolean?, server: string?, bin : string[]? | nil }?
 function M.package(package, opts)
     return function()
         return M.install(package, opts).with_receipt()
@@ -30,7 +30,7 @@ end
 
 ---@async
 ---@param pkg string: The luarock package to install.
----@param opts { dev: boolean, bin : string[] | nil } | nil
+---@param opts { dev: boolean?, server: string?, bin : string[]? | nil }?
 function M.install(pkg, opts)
     opts = opts or {}
     local ctx = installer.context()
@@ -40,6 +40,7 @@ function M.install(pkg, opts)
         "--tree",
         ctx.cwd:get(),
         opts.dev and "--dev" or vim.NIL,
+        opts.server and ("--server=%s"):format(opts.server) or vim.NIL,
         pkg,
         ctx.requested_version:or_else(vim.NIL),
     }
diff --git a/lua/mason-registry/index.lua b/lua/mason-registry/index.lua
index ccc054d..279d269 100644
--- a/lua/mason-registry/index.lua
+++ b/lua/mason-registry/index.lua
@@ -84,6 +84,7 @@ return {
   ["ltex-ls"] = "mason-registry.ltex-ls",
   ["lua-language-server"] = "mason-registry.lua-language-server",
   luacheck = "mason-registry.luacheck",
+  luaformatter = "mason-registry.luaformatter",
   marksman = "mason-registry.marksman",
   ["metamath-zero-lsp"] = "mason-registry.metamath-zero-lsp",
   misspell = "mason-registry.misspell",
diff --git a/lua/mason-registry/luaformatter/init.lua b/lua/mason-registry/luaformatter/init.lua
index aae4ef7..f4e4c88 100644
--- a/lua/mason-registry/luaformatter/init.lua
+++ b/lua/mason-registry/luaformatter/init.lua
@@ -1,8 +1,5 @@
 local Pkg = require "mason-core.package"
-local git = require "mason-core.managers.git"
-local github = require "mason-core.managers.github"
-local platform = require "mason-core.platform"
-local Optional = require "mason-core.optional"
+local luarocks = require "mason-core.managers.luarocks"
 
 return Pkg.new {
     name = "luaformatter",
@@ -10,14 +7,8 @@ return Pkg.new {
     homepage = "https://github.com/Koihik/LuaFormatter",
     languages = { Pkg.Lang.Lua },
     categories = { Pkg.Cat.Formatter },
-    ---@async
-    ---@param ctx InstallContext
-    install = function(ctx)
-        local source = github.tag { repo = "Koihik/LuaFormatter" }
-        source.with_receipt()
-        git.clone { "https://github.com/Koihik/LuaFormatter", version = Optional.of(source.tag), recursive=true }
-        ctx.spawn.cmake { "." }
-        ctx.spawn.make { }
-        ctx:link_bin("lua-format", platform.is.win and "lua-format.exe" or "lua-format")
-    end,
+    install = luarocks.package("luaformatter", {
+        server = "https://luarocks.org/dev",
+        bin = { "lua-format" },
+    }),
 }
diff --git a/lua/mason/mappings/language.lua b/lua/mason/mappings/language.lua
index 9ec1c63..56129d7 100644
--- a/lua/mason/mappings/language.lua
+++ b/lua/mason/mappings/language.lua
@@ -108,4 +108,4 @@ return {
   xml = { "lemminx" },
   yaml = { "prettier", "yaml-language-server" },
   zig = { "zls" }
-}
+}
\ No newline at end of file
diff --git a/tests/mason-core/managers/luarocks_spec.lua b/tests/mason-core/managers/luarocks_spec.lua
index 749a731..26ddcb4 100644
--- a/tests/mason-core/managers/luarocks_spec.lua
+++ b/tests/mason-core/managers/luarocks_spec.lua
@@ -15,6 +15,7 @@ describe("luarocks manager", function()
                 "--tree",
                 path.package_prefix "dummy",
                 vim.NIL, -- --dev flag
+                vim.NIL, -- --server flag
                 "lua-cjson",
                 vim.NIL, -- version
             }
@@ -33,6 +34,7 @@ describe("luarocks manager", function()
                 "--tree",
                 path.package_prefix "dummy",
                 vim.NIL, -- --dev flag
+                vim.NIL, -- --server flag
                 "lua-cjson",
                 "1.2.3",
             }
@@ -51,12 +53,32 @@ describe("luarocks manager", function()
                 "--tree",
                 path.package_prefix "dummy",
                 "--dev",
+                vim.NIL, -- --server flag
                 "lua-cjson",
                 vim.NIL, -- version
             }
         end)
     )
 
+    it(
+        "should provide --server flag",
+        async_test(function()
+            local handle = InstallHandleGenerator "dummy"
+            local ctx = InstallContextGenerator(handle)
+            installer.run_installer(ctx, luarocks.package("luaformatter", { server = "https://luarocks.org/dev" }))
+            assert.spy(ctx.spawn.luarocks).was_called(1)
+            assert.spy(ctx.spawn.luarocks).was_called_with {
+                "install",
+                "--tree",
+                path.package_prefix "dummy",
+                vim.NIL, -- --dev flag
+                "--server=https://luarocks.org/dev",
+                "luaformatter",
+                vim.NIL, -- version
+            }
+        end)
+    )
+
     it("should parse outdated luarocks", function()
         assert.same(
             {
```

**Comment by williamboman** [comment]
@williambotman /apply

```diff
diff --git a/lua/mason-core/managers/luarocks/init.lua b/lua/mason-core/managers/luarocks/init.lua
index 0deb2e1..9fd59ca 100644
--- a/lua/mason-core/managers/luarocks/init.lua
+++ b/lua/mason-core/managers/luarocks/init.lua
@@ -21,7 +21,7 @@ local function with_receipt(package)
 end
 
 ---@param package string: The luarock package to install.
----@param opts { dev: boolean, bin : string[] | nil } | nil
+---@param opts { dev: boolean?, server: string?, bin : string[]? | nil }?
 function M.package(package, opts)
     return function()
         return M.install(package, opts).with_receipt()
@@ -30,7 +30,7 @@ end
 
 ---@async
 ---@param pkg string: The luarock package to install.
----@param opts { dev: boolean, bin : string[] | nil } | nil
+---@param opts { dev: boolean?, server: string?, bin : string[]? | nil }?
 function M.install(pkg, opts)
     opts = opts or {}
     local ctx = installer.context()
@@ -40,6 +40,7 @@ function M.install(pkg, opts)
         "--tree",
         ctx.cwd:get(),
         opts.dev and "--dev" or vim.NIL,
+        opts.server and ("--server=%s"):format(opts.server) or vim.NIL,
         pkg,
         ctx.requested_version:or_else(vim.NIL),
     }
diff --git a/lua/mason-registry/luaformatter/init.lua b/lua/mason-registry/luaformatter/init.lua
new file mode 100644
index 0000000..f4e4c88
--- /dev/null
+++ b/lua/mason-registry/luaformatter/init.lua
@@ -0,0 +1,14 @@
+local Pkg = require "mason-core.package"
+local luarocks = require "mason-core.managers.luarocks"
+
+return Pkg.new {
+    name = "luaformatter",
+    desc = [[Code formatter for Lua]],
+    homepage = "https://github.com/Koihik/LuaFormatter",
+    languages = { Pkg.Lang.Lua },
+    categories = { Pkg.Cat.Formatter },
+    install = luarocks.package("luaformatter", {
+        server = "https://luarocks.org/dev",
+        bin = { "lua-format" },
+    }),
+}
diff --git a/tests/mason-core/managers/luarocks_spec.lua b/tests/mason-core/managers/luarocks_spec.lua
index 749a731..26ddcb4 100644
--- a/tests/mason-core/managers/luarocks_spec.lua
+++ b/tests/mason-core/managers/luarocks_spec.lua
@@ -15,6 +15,7 @@ describe("luarocks manager", function()
                 "--tree",
                 path.package_prefix "dummy",
                 vim.NIL, -- --dev flag
+                vim.NIL, -- --server flag
                 "lua-cjson",
                 vim.NIL, -- version
             }
@@ -33,6 +34,7 @@ describe("luarocks manager", function()
                 "--tree",
                 path.package_prefix "dummy",
                 vim.NIL, -- --dev flag
+                vim.NIL, -- --server flag
                 "lua-cjson",
                 "1.2.3",
             }
@@ -51,12 +53,32 @@ describe("luarocks manager", function()
                 "--tree",
                 path.package_prefix "dummy",
                 "--dev",
+                vim.NIL, -- --server flag
                 "lua-cjson",
                 vim.NIL, -- version
             }
         end)
     )
 
+    it(
+        "should provide --server flag",
+        async_test(function()
+            local handle = InstallHandleGenerator "dummy"
+            local ctx = InstallContextGenerator(handle)
+            installer.run_installer(ctx, luarocks.package("luaformatter", { server = "https://luarocks.org/dev" }))
+            assert.spy(ctx.spawn.luarocks).was_called(1)
+            assert.spy(ctx.spawn.luarocks).was_called_with {
+                "install",
+                "--tree",
+                path.package_prefix "dummy",
+                vim.NIL, -- --dev flag
+                "--server=https://luarocks.org/dev",
+                "luaformatter",
+                vim.NIL, -- version
+            }
+        end)
+    )
+
     it("should parse outdated luarocks", function()
         assert.same(
             {
```

**Comment by williamboman** [comment]
@williambotman /apply

```diff
diff --git a/lua/mason-core/managers/luarocks/init.lua b/lua/mason-core/managers/luarocks/init.lua
index 0deb2e1..9fd59ca 100644
--- a/lua/mason-core/managers/luarocks/init.lua
+++ b/lua/mason-core/managers/luarocks/init.lua
@@ -21,7 +21,7 @@ local function with_receipt(package)
 end
 
 ---@param package string: The luarock package to install.
----@param opts { dev: boolean, bin : string[] | nil } | nil
+---@param opts { dev: boolean?, server: string?, bin : string[]? | nil }?
 function M.package(package, opts)
     return function()
         return M.install(package, opts).with_receipt()
@@ -30,7 +30,7 @@ end
 
 ---@async
 ---@param pkg string: The luarock package to install.
----@param opts { dev: boolean, bin : string[] | nil } | nil
+---@param opts { dev: boolean?, server: string?, bin : string[]? | nil }?
 function M.install(pkg, opts)
     opts = opts or {}
     local ctx = installer.context()
@@ -40,6 +40,7 @@ function M.install(pkg, opts)
         "--tree",
         ctx.cwd:get(),
         opts.dev and "--dev" or vim.NIL,
+        opts.server and ("--server=%s"):format(opts.server) or vim.NIL,
         pkg,
         ctx.requested_version:or_else(vim.NIL),
     }
diff --git a/lua/mason-registry/index.lua b/lua/mason-registry/index.lua
index 22419eb..d65d1b5 100644
--- a/lua/mason-registry/index.lua
+++ b/lua/mason-registry/index.lua
@@ -85,6 +85,7 @@ return {
   ["lua-language-server"] = "mason-registry.lua-language-server",
   luacheck = "mason-registry.luacheck",
   markdownlint = "mason-registry.markdownlint",
+  luaformatter = "mason-registry.luaformatter",
   marksman = "mason-registry.marksman",
   ["metamath-zero-lsp"] = "mason-registry.metamath-zero-lsp",
   misspell = "mason-registry.misspell",
@@ -158,4 +159,4 @@ return {
   ["yaml-language-server"] = "mason-registry.yaml-language-server",
   zk = "mason-registry.zk",
   zls = "mason-registry.zls"
-}
\ No newline at end of file
+}
diff --git a/lua/mason-registry/luaformatter/init.lua b/lua/mason-registry/luaformatter/init.lua
index aae4ef7..dd9243d 100644
--- a/lua/mason-registry/luaformatter/init.lua
+++ b/lua/mason-registry/luaformatter/init.lua
@@ -1,8 +1,5 @@
 local Pkg = require "mason-core.package"
-local git = require "mason-core.managers.git"
-local github = require "mason-core.managers.github"
-local platform = require "mason-core.platform"
-local Optional = require "mason-core.optional"
+local luarocks = require "mason-core.managers.luarocks"
 
 return Pkg.new {
     name = "luaformatter",
@@ -10,14 +7,7 @@ return Pkg.new {
     homepage = "https://github.com/Koihik/LuaFormatter",
     languages = { Pkg.Lang.Lua },
     categories = { Pkg.Cat.Formatter },
-    ---@async
-    ---@param ctx InstallContext
-    install = function(ctx)
-        local source = github.tag { repo = "Koihik/LuaFormatter" }
-        source.with_receipt()
-        git.clone { "https://github.com/Koihik/LuaFormatter", version = Optional.of(source.tag), recursive=true }
-        ctx.spawn.cmake { "." }
-        ctx.spawn.make { }
-        ctx:link_bin("lua-format", platform.is.win and "lua-format.exe" or "lua-format")
-    end,
+    install = luarocks.package("luaformatter", {
+        server = "https://luarocks.org/dev",
+    }),
 }
diff --git a/lua/mason/mappings/language.lua b/lua/mason/mappings/language.lua
index 74227a3..dc8beab 100644
--- a/lua/mason/mappings/language.lua
+++ b/lua/mason/mappings/language.lua
@@ -108,4 +108,4 @@ return {
   xml = { "lemminx" },
   yaml = { "prettier", "yaml-language-server" },
   zig = { "zls" }
-}
+}
\ No newline at end of file
```

**Comment by williamboman** [comment]
Superseded by #116, thanks! I tried to patch this PR directly but things get problematic when the source branch is the same as the default branch (`alpha`)

**Pull request closed without merging**